src/Features/CSharp/Portable/CodeRefactorings/NamedArguments/CSharpAddNamedArgumentsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,188 @@
+            }
+
+            for (int index = 0; index < arguments.Count; ++index)
+            {
+                var parameter = parameters[index];
+                var argument = arguments[index];
+
+                if (argument.NameColon != null)
+                {
+                    yield return argument;
+                }
+                else if (parameter.IsParams && arguments.Count - index > 1)
+                {
+                    var expressions = arguments.Skip(index).Select(arg => arg.Expression);
+                    var arrayCreation = WrapInArrayCreation(expressions);
+                    var namedArgument = Argument(arrayCreation).WithNameColon(NameColon(parameter.Name));
+                    yield return namedArgument;
+                    yield break;
+                }
+                else
+                {
+                    yield return argument.WithNameColon(NameColon(parameter.Name));
+                }
+            }
+        }
+
+        private static ImplicitArrayCreationExpressionSyntax WrapInArrayCreation(IEnumerable<ExpressionSyntax> expressions)
+        {
+            var initializer = InitializerExpression(SyntaxKind.ArrayInitializerExpression, SeparatedList(expressions));
+            return ImplicitArrayCreationExpression(initializer);
+        }
+
+        protected override bool IsCandidate(SyntaxNode node)
+        {
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            switch (node.Kind())
+            {
+                case SyntaxKind.InvocationExpression:
+                case SyntaxKind.ElementAccessExpression:
+                case SyntaxKind.ObjectCreationExpression:
+                case SyntaxKind.BaseConstructorInitializer:
+                case SyntaxKind.ThisConstructorInitializer:
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        protected override bool TryGetArguments(SyntaxNode node, SemanticModel semanticModel, out IList<ArgumentSyntax> arguments)
+        {
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            if (semanticModel == null)
+            {
+                throw new ArgumentNullException(nameof(semanticModel));
+            }
+
+            switch (node.Kind())
+            {
+                case SyntaxKind.InvocationExpression:
+                    {
+                        var syntaxNode = (InvocationExpressionSyntax)node;
+
+                        // Skip nameof
+                        var symbol = semanticModel.GetSymbolInfo(syntaxNode).Symbol;
+                        if (symbol == null)
+                        {
+                            goto default;
+                        }
+
+                        arguments = syntaxNode.ArgumentList.Arguments.ToList();
```

Comment:
Will do. Regarding collection types, I personally don't feel right about this,
```cs
IEnumerable<T> GetNamedArguments(IList<ArgumentSyntax>, ImmutableArray<IParameterSymbol>)
```
Is there any convention here that I could follow?